src/phoenix/experimental/evals/functions/binary.py

```diff
@@ -58,4 +51,80 @@ def llm_eval_binary(
+        (rail_class if (rail_class := resp.strip()) in rail_classes else None) for resp in responses
+    ]
+
+
+def run_relevance_eval(
+    dataframe: pd.DataFrame,
+    query_column_name: str = "attributes.input.value",
+    retrieved_documents_column_name: str = "attributes.retrieval.documents",
+    template: str = RAG_RELEVANCY_PROMPT_TEMPLATE_STR,
+    model: Optional[BaseEvalModel] = None,
+) -> List[List[Optional[bool]]]:
+    """Given a pandas dataframe containing queries and retrieved documents,
+       classifies the relevance of each retrieved document to the corresponding
+       query using an LLM.
+
+    Args:
+        dataframe (pd.DataFrame): A pandas dataframe containing queries and
+        retrieved documents.
+
+        query_column_name (str, optional): The name of the column containing the
+        queries.
+
+        retrieved_documents_column_name (str, optional): The name of the column
+        containing the retrieved document data. Each entry in this column should be
+        a list of dictionaries containing metadata about the retrieved documents.
+
+    Returns:
+        List[List[str]]: A list of relevant and not relevant classifications.
+        The "shape" of the list should mirror the "shape" of the retrieved
+        documents column, in the sense that it has the same length as the input
+        dataframe and each sub-list has the same length as the corresponding
+        list in the retrieved documents column. The values in the sub-lists are
+        either booleans or None in the case where the LLM output could not be
+        parsed.
+    """
+
+    llm_relevance_column_name = "llm_relevance"
+    retrieved_document_text_column_name = "retrieved_document_text"
+    retrievals_df = dataframe[[query_column_name, retrieved_documents_column_name]].copy()
+    non_null_df = retrievals_df.dropna(how="any")
+    non_null_df[retrieved_documents_column_name] = non_null_df[retrieved_documents_column_name].map(
+        list
+    )
+    exploded_df = non_null_df.explode(retrieved_documents_column_name, ignore_index=False)
+    exploded_df[retrieved_document_text_column_name] = [
+        document_data["document.content"] if document_data is not None else None
+        for document_data in exploded_df[retrieved_documents_column_name]
+    ]
+    exploded_df = exploded_df.rename(
+        columns={
+            query_column_name: "query",
+            retrieved_document_text_column_name: "reference",
+        }
+    )
+    exploded_df[llm_relevance_column_name] = [
+        {"relevant": True, "irrelevant": False}.get(relevance_class)
+        if relevance_class is not None
+        else None
+        for relevance_class in llm_eval_binary(
+            exploded_df,
+            template=PromptTemplate(RAG_RELEVANCY_PROMPT_TEMPLATE_STR),
+            model=model or OpenAiModel(),
+            rails=["relevant", "irrelevant"],
+        )
+    ]
+    imploded_df = exploded_df.groupby(exploded_df.index, axis="index").agg(
```

Comment:
not sure the reverse of explode is implode...

Reply:
`collapsed_df`?